File: EstherCharLib/scan-char-assets.mjs
#!/usr/bin/env node
/**
 * Node.js 版本的角色资产扫描脚本
 * 扫描 X:\projects2024\ETH\Asset\Cha 下所有角色目录
 */

import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const CHA_ROOT = "X:\\projects2024\\ETH\\Asset\\Cha";
const OUTPUT_JSON = path.join(__dirname, "esther-char-db.json");
const OUTPUT_JS = path.join(__dirname, "esther-char-db.data.js");
const CGTW_NAMES_FILE = path.join(__dirname, "cgtw-asset-names.json");

const IMAGE_EXT = new Set([".png", ".jpg", ".jpeg", ".webp", ".bmp", ".gif", ".tga", ".tif", ".tiff", ".psd", ".exr"]);
const VIDEO_EXT = new Set([".mp4", ".mov", ".avi", ".wmv", ".webm", ".mkv", ".flv"]);
const MAYA_EXT = new Set([".ma", ".mb"]);
const FBX_EXT = new Set([".fbx"]);
const EXCLUDE_DIRS = new Set(["history", "metadata", "backup"]);

function walkDir(dir) {
  const results = [];
  if (!fs.existsSync(dir) || !fs.statSync(dir).isDirectory()) return results;
  const entries = fs.readdirSync(dir, { withFileTypes: true });
  for (const entry of entries) {
    const fullPath = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      if (EXCLUDE_DIRS.has(entry.name.toLowerCase())) continue;
      results.push(...walkDir(fullPath));
    } else if (entry.isFile()) {
      results.push(fullPath);
    }
  }
  return results.sort();
}

function collectMedia(folder, extensions) {
  return walkDir(folder)
    .filter(f => extensions.has(path.extname(f).toLowerCase()))
    .map(f => f.replace(/\\/g, "/"));
}

function collectMayaFiles(folder) {
  return collectMedia(folder, MAYA_EXT);
}

function findMainMaya(folder, charId, suffix) {
  const files = collectMayaFiles(folder);
  if (files.length === 0) return "";
  const target = `${charId}_${suffix}.ma`.toLowerCase();
  for (const f of files) {
    if (path.basename(f).toLowerCase() === target) return f;
  }
  return files[0];
}

function scanCharacter(charDir) {
  const charId = path.basename(charDir);

  const prodesignGameImages = collectMedia(path.join(charDir, "Prodesign", "work"), IMAGE_EXT);
  const prodesignFilmImages = collectMedia(path.join(charDir, "Prodesign", "approved"), IMAGE_EXT);
  let lookdevVideos = collectMedia(path.join(charDir, "LookDev", "approved"), VIDEO_EXT);
  if (lookdevVideos.length > 1) {
    let latest = lookdevVideos[0];
    let latestMtime = fs.statSync(lookdevVideos[0].replace(/\//g, path.sep)).mtimeMs;
    for (let i = 1; i < lookdevVideos.length; i++) {
      const mt = fs.statSync(lookdevVideos[i].replace(/\//g, path.sep)).mtimeMs;
      if (mt > latestMtime) { latestMtime = mt; latest = lookdevVideos[i]; }
    }
    lookdevVideos = [latest];
  }
  const lookdevImages = collectMedia(path.join(charDir, "LookDev", "approved"), IMAGE_EXT);

  const modPath = findMainMaya(path.join(charDir, "MOD", "work"), charId, "mod");
  const texPath = findMainMaya(path.join(charDir, "TEX", "publish", "maya"), charId, "tex");
  const rigPath = findMainMaya(path.join(charDir, "RIG", "publish"), charId, "rig");

  const allFbxCandidates = [
    ...walkDir(path.join(charDir, "MOD", "work")),
    ...walkDir(path.join(charDir, "RIG", "work")),
  ].filter(f => FBX_EXT.has(path.extname(f).toLowerCase()));

  let allFbx = [];
  if (allFbxCandidates.length > 0) {
    let latest = allFbxCandidates[0];
    let latestMtime = fs.statSync(allFbxCandidates[0]).mtimeMs;
    for (let i = 1; i < allFbxCandidates.length; i++) {
      const mt = fs.statSync(allFbxCandidates[i]).mtimeMs;
      if (mt > latestMtime) { latestMtime = mt; latest = allFbxCandidates[i]; }
    }
    allFbx = [latest.replace(/\\/g, "/")];
  }

  return {
    id: charId,
    name: charId,
    basePath: charDir.replace(/\\/g, "/"),
    prodesign_game: { label: "2D游戏设计", images: prodesignGameImages },
    prodesign_film: { label: "2D影视设计", images: prodesignFilmImages },
    lookdev: { label: "3D影视设计（LookDev）", videos: lookdevVideos, images: lookdevImages },
    mod: { label: "模型", path: modPath, fbxFiles: allFbx },
    tex: { label: "材质", path: texPath },
    rig: { label: "绑定", path: rigPath },
  };
}

function loadCgtwNames() {
  if (!fs.existsSync(CGTW_NAMES_FILE)) return new Map();
  const data = JSON.parse(fs.readFileSync(CGTW_NAMES_FILE, "utf-8"));
  const map = new Map();
  for (const item of data) {
    map.set(item.entity, item);
  }
  return map;
}

function main() {
  if (!fs.existsSync(CHA_ROOT)) {
    console.log(`根目录不存在: ${CHA_ROOT}`);
    return;
  }

  const cgtwMap = loadCgtwNames();
  console.log(`CGTW 名称数据: ${cgtwMap.size} 条`);

  const charDirs = fs.readdirSync(CHA_ROOT, { withFileTypes: true })
    .filter(d => d.isDirectory())
    .map(d => path.join(CHA_ROOT, d.name))
    .sort();

  console.log(`扫描到 ${charDirs.length} 个角色目录`);

  const characters = [];
  for (const charDir of charDirs) {
    const charId = path.basename(charDir);
    process.stdout.write(`  扫描: ${charId} ...`);
    const charData = scanCharacter(charDir);

    const cgtw = cgtwMap.get(charId);
    if (!cgtw || !cgtw.notes) {
      console.log(` [跳过 - 无自定义备注]`);
      continue;
    }

    charData.cn_name = cgtw.cn_name || "";
    charData.first_appear = cgtw.first_appear || "";
    charData.create_time = cgtw.create_time || "";
    charData.notes_episode = cgtw.notes_episode || "";
    charData.notes_cn = cgtw.notes_cn || "";
    charData.notes_en = cgtw.notes_en || "";
    charData.notes_role = cgtw.notes_role || "";

    characters.push(charData);

    const counts = [];
    const pg = charData.prodesign_game.images.length;
    const pf = charData.prodesign_film.images.length;
    const lv = charData.lookdev.videos.length;
    const li = charData.lookdev.images.length;
    const fbx = (charData.mod.fbxFiles || []).length;
    if (pg) counts.push(`2D游戏:${pg}`);
    if (pf) counts.push(`2D影视:${pf}`);
    if (lv) counts.push(`LookDev视频:${lv}`);
    if (li) counts.push(`LookDev图:${li}`);
    if (charData.mod.path) counts.push("MOD");
    if (fbx) counts.push(`FBX:${fbx}`);
    if (charData.tex.path) counts.push("TEX");
    if (charData.rig.path) counts.push("RIG");
    console.log(` [${counts.length ? counts.join(", ") : "暂无资源"}]`);
  }

  const db = {
    version: "1.0.0",
    project: "伊瑟 ETH",
    source: CHA_ROOT.replace(/\\/g, "/"),
    generatedAt: new Date().toISOString(),
    totalCharacters: characters.length,
    characters,
  };

  fs.writeFileSync(OUTPUT_JSON, JSON.stringify(db, null, 2), "utf-8");
  fs.writeFileSync(OUTPUT_JS, "window.__ESTHER_CHAR_DB__ = " + JSON.stringify(db, null, 2) + ";\n", "utf-8");

  console.log(`\n完成！共 ${characters.length} 个角色`);
  console.log(`JSON 数据库: ${OUTPUT_JSON}`);
  console.log(`JS 数据库:   ${OUTPUT_JS}`);
}

main();

File: TorchCharLib/scan-char-assets.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
扫描 X:\projects2024\TLI\Asset\Cha 下所有角色目录，
按照统一资产结构生成 torch-char-db.json
"""

import json
import os
from pathlib import Path
from datetime import datetime, timezone

CHA_ROOT = r"X:\projects2024\TLI\Asset\Cha"
OUTPUT_JSON = os.path.join(os.path.dirname(os.path.abspath(__file__)), "torch-char-db.json")
OUTPUT_JS = os.path.join(os.path.dirname(os.path.abspath(__file__)), "torch-char-db.data.js")

IMAGE_EXT = {".png", ".jpg", ".jpeg", ".webp", ".bmp", ".gif", ".tga", ".tif", ".tiff", ".psd", ".exr"}
VIDEO_EXT = {".mp4", ".mov", ".avi", ".wmv", ".webm", ".mkv", ".flv"}
MAYA_EXT = {".ma", ".mb"}

EXCLUDE_DIRS = {"history", "metadata", "backup"}


def collect_media(folder: Path, extensions: set):
    """收集目录下的媒体文件（排除 history/metadata/backup 子目录）"""
    if not folder.exists() or not folder.is_dir():
        return []
    results = []
    for f in sorted(folder.rglob("*")):
        if not f.is_file():
            continue
        if any(part.lower() in EXCLUDE_DIRS for part in f.relative_to(folder).parts):
            continue
        if f.suffix.lower() in extensions:
            results.append(str(f).replace("\\", "/"))
    return results


def collect_maya_files(folder: Path):
    """收集目录下的 Maya 文件（排除 history/metadata/backup）"""
    if not folder.exists() or not folder.is_dir():
        return []
    results = []
    for f in sorted(folder.rglob("*")):
        if not f.is_file():
            continue
        if any(part.lower() in EXCLUDE_DIRS for part in f.relative_to(folder).parts):
            continue
        if f.suffix.lower() in MAYA_EXT:
            results.append(str(f).replace("\\", "/"))
    return results


def find_main_maya(folder: Path, char_id: str, suffix: str):
    """在目录下找主 Maya 文件，优先匹配 {char_id}_{suffix}.ma"""
    files = collect_maya_files(folder)
    if not files:
        return ""
    target = f"{char_id}_{suffix}.ma".lower()
    for f in files:
        if f.split("/")[-1].lower() == target:
            return f
    return files[0]


def scan_character(char_dir: Path):
    char_id = char_dir.name

    prodesign_game_images = collect_media(char_dir / "Prodesign" / "work", IMAGE_EXT)
    prodesign_film_images = collect_media(char_dir / "Prodesign" / "approved", IMAGE_EXT)
    game_3d_images = []  # 路径待定
    lookdev_videos = collect_media(char_dir / "LookDev" / "approved", VIDEO_EXT)
    if len(lookdev_videos) > 1:
        latest = max(lookdev_videos, key=lambda f: Path(f).stat().st_mtime)
        lookdev_videos = [latest]
    lookdev_images = collect_media(char_dir / "LookDev" / "approved", IMAGE_EXT)

    mod_path = find_main_maya(char_dir / "MOD" / "work", char_id, "mod")
    tex_path = find_main_maya(char_dir / "TEX" / "publish" / "maya", char_id, "tex")
    rig_path = find_main_maya(char_dir / "RIG" / "publish", char_id, "rig")

    return {
        "id": char_id,
        "name": char_id,
        "basePath": str(char_dir).replace("\\", "/"),
        "prodesign_game": {
            "label": "2D游戏设计",
            "images": prodesign_game_images,
        },
        "prodesign_film": {
            "label": "2D影视设计",
            "images": prodesign_film_images,
        },
        "game_3d": {
            "label": "3D游戏设计",
            "images": game_3d_images,
        },
        "lookdev": {
            "label": "3D影视设计（LookDev）",
            "videos": lookdev_videos,
            "images": lookdev_images,
        },
        "mod": {
            "label": "模型",
            "path": mod_path,
        },
        "tex": {
            "label": "贴图 Maya",
            "path": tex_path,
        },
        "rig": {
            "label": "绑定",
            "path": rig_path,
        },
    }


def main():
    cha_root = Path(CHA_ROOT)
    if not cha_root.exists():
        print(f"根目录不存在: {CHA_ROOT}")
        return

    char_dirs = sorted([d for d in cha_root.iterdir() if d.is_dir()])
    print(f"扫描到 {len(char_dirs)} 个角色目录")

    characters = []
    for d in char_dirs:
        print(f"  扫描: {d.name} ...", end="")
        char_data = scan_character(d)
        characters.append(char_data)

        counts = []
        pg = len(char_data["prodesign_game"]["images"])
        pf = len(char_data["prodesign_film"]["images"])
        g3 = len(char_data["game_3d"]["images"])
        lv = len(char_data["lookdev"]["videos"])
        li = len(char_data["lookdev"]["images"])
        if pg: counts.append(f"2D游戏:{pg}")
        if pf: counts.append(f"2D影视:{pf}")
        if g3: counts.append(f"3D游戏:{g3}")
        if lv: counts.append(f"LookDev视频:{lv}")
        if li: counts.append(f"LookDev图:{li}")
        if char_data["mod"]["path"]: counts.append("MOD")
        if char_data["tex"]["path"]: counts.append("TEX")
        if char_data["rig"]["path"]: counts.append("RIG")
        print(f" [{', '.join(counts) if counts else '暂无资源'}]")

    db = {
        "version": "1.0.0",
        "project": "火炬 TLI",
        "source": CHA_ROOT.replace("\\", "/"),
        "generatedAt": datetime.now(timezone.utc).isoformat(),
        "totalCharacters": len(characters),
        "characters": characters,
    }

    with open(OUTPUT_JSON, "w", encoding="utf-8") as f:
        json.dump(db, f, ensure_ascii=False, indent=2)

    with open(OUTPUT_JS, "w", encoding="utf-8") as f:
        f.write("window.__TORCH_CHAR_DB__ = ")
        json.dump(db, f, ensure_ascii=False, indent=2)
        f.write(";\n")

    print(f"\n完成！共 {len(characters)} 个角色")
    print(f"JSON 数据库: {OUTPUT_JSON}")
    print(f"JS 数据库:   {OUTPUT_JS}")


if __name__ == "__main__":
    main()
